### DISCUSSION TITLE
OpenSimulator Is Here !

### DISCUSSION THREAD

**Discussion by AdilElFarissi** [btcpayserver/btcpayserver#5384] [Plugins & integrations]
Hi there.
My name is Adil El Farissi (48yo), I am an economist, a social entrepreneur and autodidact coder since 2003, but above all I am an eternal amateur in action and the best slave of my muse... that's why I do this kind of thing :)

I am happy to announce the upcoming publication of my modest contribution to the BTCPay saga with a new use case and a plugin allowing a secure interaction between LSL scripted objects into the OpenSimulator virtual worlds (the cubes in the video) and a BTCPay instance and store to operate all kinds of transactions using the BTCPay invoicing system. The following video is a proof of work and concept (expand the descriptions for links and infos):
https://www.youtube.com/watch?v=NTb8PPp0qco

Unfortunately, for several technical and security reasons, the OpenSimulator virtual worlds must be considered as a vulnerable environment in which sensitive data such as the account or Greenfield API keys cannot be transmitted or stored in a safe way. For this the plugin use 2 public endpoints, one for the authorizations processing and the other is the same as the Pay Button to handle the invoicing part... the difference is that the plugin returns Json instead of views and only allows the invoices creation to the  objects previously authorized by the store owner under the "Authorization Manager" section.

I wrote the scripts under the "LSL Scripts" section in a way that allows the virtual worlds scripters to easily understand how it works by dividing the script into 2 parts (states) like our endpoints, one part manages the authorizations and the other the invoicing... and just by copying/pasting blocks of code, they will be able to recompose functional scripts with less code and adapted to their needs. In the on-board documentation, I explained in detail how the plugin and the scripts interact and I think this is enough material for any scripter.

I was going to publish a pre-release today but fortunately or unfortunately, I found a way to get the rates without using the API key (Newbie XD), so I decided to add a step before the invoice request in which the customer will have a blue box that displays an estimate of the amount to pay in cryptos if the currency is fiat and asks him if he has enough coins to honor the transaction. This will avoid creating an invoice for nothing in the case where the customer does not have enough coins...

I would like to have informations about the peers review and audits... is it enough to use the plugin builder or is there a process to follow? I will let you know when the plugin is ready right here first. 

If you have questions, feel free to ask. Thank you and keep up the good work :)

**Comment by dennisreimann** [comment]
Hey, first of all congrats for reaching this point, this looks like a lot of tinkering and learning stuff along the way! :)

> is it enough to use the plugin builder or is there a process to follow?

You can register at the plugin builder, add your plugin and then do builds and publish them as releases. There is no audit step in  between, people will have to verify what they are installing/running plugin-wise.

How was the process of building the plugina nd did you encounter steps that could be simplified or made more clear in the docs? Let us know if there's something you were/are struggling with.

**Reply by AdilElFarissi** [reply]
> Hey, first of all congrats for reaching this point, this looks like a lot of tinkering and learning stuff along the way! :)

Thank you for your kind words, for me it doesn't matter how long the way is if it is full of fun and learning and at the end, a viable and secure solution for accepting cryptocurrency payments in the virtual worlds which will be useful for thousands of people.

Please don't be impressed by all of this because it's very simple to do... it's just this is maybe the first time you know about OpenSimulator and HyperGrid. this is something that I invite you to discover because it is easy to create the BTCPay Virtual Academy using OpenSimulator and existing content... in 2D it is comparable to installing a CMS and a template.

>There is no audit step in between, people will have to verify what they are installing/running plugin-wise.

The developer in me says: this is a bad practice and may lead to dramas...
The social entrepreneur in me says: this is an opportunity to create jobs for your team by setting up an audit and labeling service which will allow end users to have a minimum of security and confidence in the plugins knowing that your team has reviewed the code...

>How was the process of building the plugina nd did you encounter steps that could be simplified or made more clear in the docs? Let us know if there's something you were/are struggling with.

I used the template plugin as base and i will never be able to explain how it is possible that you and Mr Dorier had written exactly and by exactly i mean that even the menu icon that i had in mind to use you added it!. I only had to change a few names and add the external and internal controllers, views and what was needed for the database... even the plugin can be considered as a modified version of the Pay Button with a permissions layer...

I followed the instructions in the documentation but unfortunately the build of the plugin and the sub-module failed... after several attempts i decided to port the plugin into BTCPay and i manually placed everything in its context and it worked. I started working on it directly in BTCPay and even now it's still there in 1.11.5...

I did the opposite to prepare the plugin for the plugin builder... I wanted to test the publication of a pre-release and install it on 1.11.6 but I'm not sure that it will work... I will test this after the modifications on the 3 scripts and a final full test.

Thank you and have a nice day :)

**Comment by AdilElFarissi** [comment]
I am happy to inform you about the immediate availability of the OpenSimulator Plugin in the pre-release phase.

I added in the included documentation the quickest method to configure a local sandbox and all the LSL and OSSL functions used in the included scripts are linked to their documentation to make it easier for you to understand the LSL scripts... it's very easy for developers at your level.

If the BTCPay team have a bit of time, please review my modest amateur in action work and let me know if there is something that may harm the good work of the BTCPay instance where is installed. i will click the release button in the plugin builder only when i will have your green light.

Known bugs:
<s>The conversion of decimal to string may return the amounts with a coma "0,00" and not a dot "0.00" in non-english operaiting systems. In this case BTCPay don't send the notifications needed by the scripts to work...</s>
https://github.com/btcpayserver/btcpayserver/issues/5361 Fixed in 1.11.7

Thank you & Have Fun ^^

**Reply by AdilElFarissi** [reply]
Thank you for the fix and mention :)
I updated and installed the plugin from the plugin-builder and everything went well. I' will work on the video tutorials for the plugin installation and scripts guides this weekend...

Thank you again and have a nice WE :)

**Comment by AdilElFarissi** [comment]
Hi there, the amateur in action hits again XD.

I would just like to inform you about the progress of my modest R&D around OpenSimulator and BTCPay integration... i made a second video regarding the sale of local currencies using BTCPay for grid operators...
In this method, we use a PHP background which is much more secure than LSL scripts and allows the use of the Greenfield API and a webhook... so, this is the first working thing...
https://www.youtube.com/watch?v=x7I-mgH-p0c

The next, is the implementation of the withdrawals (local currencies buyback) using the Pull Payment system... and after, i will pluginify the server part of DTL/NSL and make the region module part talk in json instead of xml-rpc...

Thank you & @++

**Comment by AdilElFarissi** [comment]
Hi there :) 
sorry if I'm not too present... i finally found a bit of time to work on this integration... it allows to sell the local currency using BTCPay invoices and buyback/withdraw crypto using Pull Payments from a virtual world under OpenSimulator... i will do an other proof of work video to add the web interfaces and the withdrawal process...
https://github.com/AdilElFarissi/nsl-opensim-currency-for-btcpay

About my 2 cents plugin, i had a some testers... so, nobody died or complained... i released it .
Thank you and good continuation :)
